lib/Frontend/CompilerInvocation.cpp

```diff
@@ -206,28 +195,27 @@ class ArgsToFrontendInputsConverter {
     }
     // Keep buffer alive because code passes around StringRefs.
     FilelistBuffer = std::move(*filelistBufferOrError);
+    bool hadDuplicates = false;
     for (auto file : llvm::make_range(llvm::line_iterator(*FilelistBuffer),
-                                      llvm::line_iterator())) {
-      addFile(file, Whence::SecondaryFromFileList);
-    }
-    return false;
+                                      llvm::line_iterator()))
+      hadDuplicates = addFile(file) || hadDuplicates;
+    return false; // Don't bail out for duplicates, too many tests depend on it.
   }
 
-  bool setPrimaryFiles() {
-    for (StringRef primaryFile : PrimaryFiles) {
-      const auto iterator = FileIndices.find(primaryFile);
-      // Catch "swiftc -frontend -c -filelist foo -primary-file
-      // some-file-not-in-foo".
-      if (iterator == FileIndices.end()) {
-        assert(FilelistPathArg != nullptr &&
-               "Missing primary with no filelist");
-        Diags.diagnose(SourceLoc(), diag::error_primary_file_not_found,
-                       primaryFile, filelistPath());
-        return true;
-      }
-      Inputs.addPrimaryInputFilename(iterator->second);
+  bool addFile(StringRef file) {
+    if (Files.count(file) == 0) {
+      Files.insert(file);
+      return false;
     }
-    return false;
+    Diags.diagnose(SourceLoc(), diag::error_duplicate_input_file, file);
+    return true;
+  }
+
+  std::set<StringRef> getPrimaries() const {
```

Comment:
`std::set` is a pretty heavy structure (see the [LLVM Programmer's Manual](http://www.llvm.org/docs/ProgrammersManual.html#set)) [`llvm::SmallSet` or `llvm::StringSet`](http://www.llvm.org/docs/ProgrammersManual.html#set-like-containers-std-set-smallset-setvector-etc) might be better here.

(This isn't super important because we really might have cases where there are a lot of primary files, and then you're dealing with some of the downsides of the other types too. But it's just one of those LLVM instincts.)

Reply:
Replace with the StringSet.